File: skipthedishes-gui/src/app/services/customer/customer.service.ts
import {Customer} from "./customer.model";
import { HttpClient} from "@angular/common/http";
import {Injectable} from "@angular/core";
import {Observable} from "rxjs/Observable";
import "rxjs/add/observable/of";
import "rxjs/add/operator/do";
import {Dish} from "../search/dish.model";
import {Restaurant} from "../search/restaurant.model";

@Injectable()
export class CustomerService{

  customers:Customer[];

  currentCustomer:Customer;

  dishes:Dish[]=[];
  restaurants:Restaurant[]=[];


  constructor(private http:HttpClient){

  }

  public loadCustomers():Observable<Customer[]>{
    if(this.customers){
      return Observable.of(this.customers);
    }

    return this.http.get<Customer[]>('/api/customers').do((result:Customer[])=>{
       this.customers = result;
    });

  }

  isCurrentCustomer(customer:Customer):boolean{
    if(this.currentCustomer && customer && this.currentCustomer.id===customer.id){
      return true;
    }

    return false;

  }

  public loadFavorites(){
    this.http.get<Dish[]>("/api/dishes",{params:{
      "customerId":this.currentCustomer.id
    }}).subscribe((result:Dish[])=>{
       this.dishes = result;
    });

    this.http.get<Restaurant[]>("/api/restaurants",{params:{
      "customerId":this.currentCustomer.id
    }}).subscribe((result:Restaurant[])=>{
      this.restaurants = result;
    });

  }

  public addFavoriteRestaurant(restaurant:Restaurant){
    this.http.put<Customer>("/api/customers/"+this.currentCustomer.id+"/add-favorite-restaurant",null,{params:{
      "restaurantId":restaurant.id
    }}).subscribe((customer:Customer)=>{
       this.currentCustomer.favoriteRestaurants = customer.favoriteRestaurants;
    })
  }

  public removeFavoriteRestaurant(restaurant:Restaurant){
    this.http.put<Customer>("/api/customers/"+this.currentCustomer.id+"/remove-favorite-restaurant",null,{params:{
      "restaurantId":restaurant.id
    }}).subscribe((customer:Customer)=>{
      this.currentCustomer.favoriteRestaurants = customer.favoriteRestaurants;
    })
  }

  public addFavoriteDish(restaurant:Dish){
    this.http.put<Customer>("/api/customers/"+this.currentCustomer.id+"/add-favorite-dish",null,{params:{
      "dishId":restaurant.id
    }}).subscribe((customer:Customer)=>{
      this.currentCustomer.favoriteDishes = customer.favoriteDishes;
    })
  }

  public removeFavoriteDish(restaurant:Dish){
    this.http.put<Customer>("/api/customers/"+this.currentCustomer.id+"/remove-favorite-dish",null,{params:{
      "dishId":restaurant.id
    }}).subscribe((customer:Customer)=>{
      this.currentCustomer.favoriteDishes = customer.favoriteDishes;
    })
  }

  updateCurrentCustomer(){
    this.http.get<Customer>("api/customers/"+this.currentCustomer.id).subscribe((result:Customer)=>{
      this.currentCustomer.dishCoin = result.dishCoin;
    });
  }

}

File: skipthedishes-api/src/main/java/com/skipthedishes/api/services/CustomerService.java
package com.skipthedishes.api.services;

import com.skipthedishes.api.entities.Customer;
import com.skipthedishes.api.entities.Dish;
import com.skipthedishes.api.entities.Restaurant;
import com.skipthedishes.api.repositories.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.stereotype.Service;

@Service
public class CustomerService {

    @Autowired
    private CustomerRepository customerRepository;

    public Customer addFavoriteDish(String customerId, String dishId) {
        Customer customerRecovered = customerRepository.findOne(customerId);
        if (customerRecovered == null) {
            throw new EmptyResultDataAccessException(1);
        }
        customerRecovered.getFavoriteDishes().add(dishId);
        return customerRepository.save(customerRecovered);
    }

    public Customer removeFavoriteDish(String customerId, String dishId) {
        Customer customerRecovered = customerRepository.findOne(customerId);
        if (customerRecovered == null) {
            throw new EmptyResultDataAccessException(1);
        }
        customerRecovered.getFavoriteDishes().remove(dishId);
        return customerRepository.save(customerRecovered);
    }

    public Customer addFavoriteRestaurant(String customerId, String restaurantId) {
        Customer customerRecovered = customerRepository.findOne(customerId);
        if (customerRecovered == null) {
            throw new EmptyResultDataAccessException(1);
        }
        customerRecovered.getFavoriteRestaurants().add(restaurantId);
        return customerRepository.save(customerRecovered);
    }

    public Customer removeFavoriteRestaurant(String customerId, String restaurantId) {
        Customer customerRecovered = customerRepository.findOne(customerId);
        if (customerRecovered == null) {
            throw new EmptyResultDataAccessException(1);
        }
        customerRecovered.getFavoriteRestaurants().remove(restaurantId);
        return customerRepository.save(customerRecovered);
    }
}
